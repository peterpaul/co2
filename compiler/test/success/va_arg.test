package compiler.test;

["stdarg.h"] {
	void va_copy(va_list dest, va_list src);
	void va_end(va_list list);
	int printf(char * fmt, ...);
}

int accumulate(int size, va_list app)
{
	return size + va_arg(app, int);
}

int map_args(int size, ...)
{
	int i;
	int result = 0;
	for (i = 0; i < size; i = i + 1) 
	{
		va_list aq;
		va_copy(aq, ...);
		result = accumulate(result, aq);
		va_end(aq);
	}
	return result;
}

int fun_va_param(int a, ...)
{
	return a + va_arg(int);
}

int main()
{
	int i, size = 15;
	int [] array = new int[size];
	for (i = 0; i < size; i = i + 1) {
		array[i] = 1;
	}
	printf("result = %d\n", map_args(5, 1, 1, 1, 1, 1));
	printf("fun_va_param(4, 5) = %d\n", fun_va_param(4,5));
	return 0;
}