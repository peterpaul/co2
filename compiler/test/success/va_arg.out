result = 5
fun_va_param(4, 5) = 9
